include/mqtt/endpoint.hpp

```diff
@@ -69,6 +71,7 @@ template <typename Socket, typename Mutex = std::mutex, template<typename...> cl
 class endpoint : public std::enable_shared_from_this<endpoint<Socket, Mutex, LockGuard, PacketIdBytes>> {
     using this_type = endpoint<Socket, Mutex, LockGuard, PacketIdBytes>;
 public:
+    using std::enable_shared_from_this<endpoint<Socket, Mutex, LockGuard, PacketIdBytes>>::shared_from_this;
```

Comment:
I ran into a problem earlier today.

The way enable_shared_from_this is used in the mqtt_cpp codebase makes it so that if someone inherits from mqtt::endpoint, then

```
Derived * p;
p->shared_from_this();
```

Returns an instance of `std::shared_ptr<mqtt::endpoint>` and not `std::shared_ptr<Derived>`

One way to solve that is to use the "Curiously Recurring Template Pattern", and have the endpoint<> class take as a template parameter the type of the most-derived type of the inheritance graph, and then pass that type to std::enable_shared_from_this<>.

Reply:
Hmm... CRTP is work well with Derived. But I couldn't a good way to support endpoint as leaf.
https://wandbox.org/permlink/2jSaOy3xkNgO9sAu
Any ideas?

Reply:
I introduced `base_selector` meta function. See https://wandbox.org/permlink/Mm74LrjbYOMq4AZo
I think that it works well.

Reply:
I think that the following approach is better:

https://wandbox.org/permlink/7r1s3XgsIxNpxtNB

No meta function required.

Reply:
BTW, the issue is not related to the PR. So it should be a separate issue.